### PR TITLE
[lang] migrate to nom 6

### PR DESCRIPTION
**Pull request by tstack**
!THIS IS NOT COMPLETE! I'm mostly looking for some high-level
feedback at the moment.  Most of the tests are passing, except
for the alias-related ones.

This change migrates the lang module from nom 4 to nom 6 (the
latest release at the time).  I tried to follow the existing
code for the most part.  I didn't want to deviate too much
so that we could still compare the two impls and not introduce
too many new bugs.  The changes I have introduced are to try
and improve error handling.  I tried to follow the advice in
this post:

  https://eyalkalderon.com/blog/nom-error-recovery/

Basically, the parse should always succeed and errors are
inserted into the AST (e.g. a bad expression is represented
as an Expr::Error in the tree).  Error messages are logged
using the existing ErrorReporter mechanism.  In addition,
when unexpected data is encountered, the parser tries to
seek forward to the next section of the pipeline (i.e. the
vertical bar).

TODO:
 * Implement aliases
 * Try to handle more error cases
 * Remove dead code
 * Double check that old parser behaviors are preserved

### CODE REVIEW

**Review by rcoh — COMMENTED** [review]
just a sidenote, I've noticed some issues that are probably worth resolving (or at least figuring out a best path to resolving here) around ambiguities / priorities parsing aggregate functions and expressions where in some contexts expression parsing took priority and made it impossible to use aggregate functions.

No reproducers on hand, but basically you'd see something like `* | json | sum(foo)` <- not literally this, obviously this works. "Error, couldn't find a function sum"`

**Comment by tstack** [comment]
I think the priority would be defined by these lines:

https://github.com/rcoh/angle-grinder/pull/148/files#diff-9184a440fdb6af807da5a90cdcd5865c62abef73df0a89c2890129ee9e9a011cR1371-R1379

The inline/aggregate operators (`inline_opers`, `multi_agg_opers`) should be tried before plain expressions `field_expr`.  So, I think it should be working, but it would be good to find an example that failed previously.  This version of the code reworked the `did_you_mean` stuff which made things a bit awkward in the previous version.

**Review by rcoh — COMMENTED** [review]
Overall this looks great. I'm very into pushing errors later into the pipeline. Way way easier than trying to get nice errors directly out of nom. And recovering to the pipe is also smart.

File: tests/structured_tests/parse_error_missing_operand.toml [hunk 1]
```diff
@@ -8,7 +8,7 @@ error = """
 error: expecting an operand for binary operator
   |
 1 | * | where response_ms *
-  |                       ^
+  |                         dangling binary operator
```
Comment (rcoh) [inline tests/structured_tests/parse_error_missing_operand.toml:11]: lost the caret here?

File: src/lang.rs [hunk 2]
```diff
@@ -185,6 +206,164 @@ impl Expr {
...
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Combinator that expects some optional whitespace followed by the vertical bar or EOF.  If that
+/// input is not found, the error message is logged.
+fn expect_bar<'a, M>(
```
Comment (rcoh) [inline src/lang.rs:360]: maybe rename to `expect_pipe` to avoid bar in the context of foo, bar, baz, etc.

File: src/lang.rs [hunk 3]
```diff
@@ -362,6 +574,208 @@ pub struct Query {
...
+        .map(|e| e.unwrap_or(Expr::Error))
+        .parse(input)
+    }
+}
+
+/// Combinator that checks for an optional keyword that should be followed by an expression
+fn kw_expr(
+    keyword: &'static str,
+    description: &'static str,
+) -> impl Clone + Fn(Span) -> IResult<Span, Option<Expr>> {
```
Comment (rcoh) [inline src/lang.rs:645]: surprised by the need for this clone bound

File: src/errors.rs [hunk 4]
```diff
@@ -1,19 +1,18 @@
...
+use std::sync::atomic::{AtomicUsize, Ordering};
 use strsim::normalized_levenshtein;
 
 /// Container for the query string that can be used to parse and report errors.
-pub struct QueryContainer {
-    query: String,
-    reporter: Box<dyn ErrorReporter>,
+pub struct QueryContainer<'a> {
+    pub query: String,
+    pub reporter: Box<dyn ErrorReporter + 'a>,
```
Comment (rcoh) [inline src/errors.rs:14]: this  `'a` is just self referential I think? I'd probably just make it `+ 'static`, there's no actual lifetime on error reporter

**Review by rcoh — COMMENTED** [review]
Overall this looks great. I love the general direction and I think it's going to make working on the parser and errors much easier.

Once we get closer to merge, there are some small things that are probably worth cleaning up. The `<xyz>_n` functions seem like a minor code smell, are they required to hook into error handling?

I also love that `do_parse` is totally gone. That was a cool piece of tech but it was so hard to debug the compiler errors.

File: tests/structured_tests/parse_error_missing_operand.toml (hunk 1 above)
Comment (tstack) [reply tests/structured_tests/parse_error_missing_operand.toml:11]: Yes, the annotate-snippets seemed to change behavior a bit and doesn't want to put the caret in when it's near the end of the snippet.  Need to fiddle with it some more.

File: src/errors.rs (hunk 4 above)
Comment (tstack) [reply src/errors.rs:14]: I think I had to add it for the unit test in the lang file.  In there, the error reporter stores strings in a vector and that vector needs to have a longer lifetime so that the output can be verified.

**Comment by tstack** [comment]
> The `<xyz>_n` functions

Sorry, I'm not sure what you're referring to here... do you mean things like `arg1` ?

**Comment by rcoh** [comment]
yeah delimited_2 delimited_3 etc.

**Comment by tstack** [comment]
> yeah delimited_2

I've renamed `delimited2` to `expect_delimited()`.  I started with `delimited2()` since it was a new version of the provided `delimited()` combinator that had some extra error handling machinery.

I also changed `arg1` to `single_arg()` since there's a a lot of operators that just take a single argument.

This latest update gets aliases working again.  I changed the alias implementation around a bit so that the alias parsing happens during static initialization instead of when it was used.

I don't really have any more plans of things to change in this pull (beyond your comments).  Hopefully, this reaches parity the the current state of the world.

File: src/lang.rs [hunk 5]
```diff
@@ -121,6 +88,58 @@ impl<T> Positioned<T> {
...
+    fn to_range(&self) -> Range<usize>;
+
+    /// Return a range from the current offset in the Span to the "sync point".  For the
+    /// angle-grinder syntax, the sync points are the vertical bar, end-of-query, and closing
+    /// braces/parens/etc...
+    fn to_sync_point(&self) -> Range<usize>;
+
+    /// Return a range from the current offset in the Span to the next whitespace
+    fn to_whitespace(&self) -> Range<usize>;
+}
```
Comment (rcoh) [inline src/lang.rs:103]: do we actually need this trait? Seems like we could just add these methods directly to Span

File: src/lang.rs [hunk 6]
```diff
@@ -383,1255 +791,1737 @@ fn is_keyword(c: char) -> bool {
...
+    alt((
+        tag("*").map(|_| ArithmeticOp::Multiply),
+        tag("/").map(|_| ArithmeticOp::Divide),
+    ))(input)
+}
 
-named!(end_of_query<Span, Span>, alt!(
-    ws!(tag!("|")) | ws!(eof!())
-));
+fn term(input: Span) -> IResult<Span, Expr> {
```
Comment (rcoh) [inline src/lang.rs:913]: can you add a short comment to this about what it parses? (and to any other parsing functions that don't have very obvious names)

File: src/lang.rs [hunk 7]
```diff
@@ -383,1255 +791,1737 @@ fn is_keyword(c: char) -> bool {
...
+        sort,
+        field_expr,
+        alias,
+        skip_to_sync,
+        did_you_mean,
+        garbage,
+    ));
+
+    separated_list1(tag("|"), opers.delimited_by(multispace0))
+        // .all_consuming()
```
Comment (rcoh) [inline src/lang.rs:1388]: delete?

File: src/lang.rs [hunk 8]
```diff
@@ -383,1255 +791,1737 @@ fn is_keyword(c: char) -> bool {
...
-                        left: Box::new(Expr::Value(data::Value::Int(4))),
-                        op: BinaryOp::Arithmetic(ArithmeticOp::Divide),
-                        right: Box::new(Expr::Value(data::Value::Int(2))),
-                    }),
-                    op: BinaryOp::Arithmetic(ArithmeticOp::Add),
-                    right: Box::new(Expr::Value(data::Value::Int(1))),
-                })
-            }
-        );
+    struct VecErrorReporter<'a> {
```
Comment (rcoh) [inline src/lang.rs:1451]: if you make this `Arc<Mutex<Vec<String>>` we can drop this lifetime + simplify the lifetimes in the trait as well

**Review by rcoh — APPROVED** [review]
LGTM! I left a couple of small cleanups inline. Thanks for doing this, this is awesome.

File: src/lang.rs (hunk 5 above)
Comment (tstack) [reply src/lang.rs:103]: I think it's needed since you can't `impl` methods on a type that is defined outside of the crate.

File: src/lang.rs (hunk 6 above)
Comment (tstack) [reply src/lang.rs:900]: added

File: src/lang.rs (hunk 8 above)
Comment (tstack) [reply src/lang.rs]: changed
